### ISSUE TITLE
Proposal: `@disableIntrinsics()` builtin function

### ISSUE THREAD

**Issue by alexrp** [ziglang/zig#22110]
`@disableIntrinsics()` forbids the compiler from recognizing code sequences and turning them into well-known library calls. The obvious examples of this would be `memcpy()`, `memset()`, etc. Additionally, it forbids the compiler from turning calls to such well-known library functions into different code (e.g. turning `memcpy()` into `__aeabi_memcpy()` on Arm).

Concretely, for LLVM, this would map to the `no-builtins` function attribute and the `nobuiltin` call site attribute, and whatever equivalent exists for other backends.

This builtin function would work similarly to `@disableInstrumentation()` in that a call to it affects the entire function body rather than the block scope. The existing `-fno-builtin` compiler option works as if `@disableIntrinsics()` is called in every function.

Consider one motivating use case from the standard library:

https://github.com/ziglang/zig/blob/aa7d138462602e086aacf738e4b92bfa3372bebe/lib/std/start.zig#L450-L488
https://github.com/ziglang/zig/blob/aa7d138462602e086aacf738e4b92bfa3372bebe/lib/std/os/linux/pie.zig#L202-L304

All of the code that runs prior to applying relocations in `std.os.linux.pie.relocate()` is potentially vulnerable to an optimization pass or compiler backend deciding to turn some code sequence into a library call, and we have to massage it just right to prevent that from happening. This is obviously quite fragile. Other kinds of software like libcs, dynamic linkers, kernels, kernel modules, etc face similar problems and consequently make use of `-fno-builtin` or [Clang's `no_builtin` attribute](https://clang.llvm.org/docs/AttributeReference.html#no-builtin).

Another motivating use case can be found here: https://github.com/ziglang/zig/issues/21831#issuecomment-2440150314

Yet more use cases would be `memcpy()`, `memset()`, etc implementations in compiler-rt.

I don't think `-fno-builtin` is an appropriate substitute for a language feature here; in the case above, we can't apply `-fno-builtin` to just that code, but we obviously don't want to apply `-fno-builtin` to the entire standard library either. That said, I don't think we need the level of granularity that Clang's `no_builtin` attribute has; I think a builtin function in the form proposed here strikes a good balance.

This proposal is related to #21833. That proposal is about preventing the compiler from turning a single call to a well-known library function into something else (e.g. `__aeabi_memcpy()` on Arm). This proposal encompasses that behavior, but only for an entire function body rather than individual calls. I suspect this will be good enough in practice, so this proposal could be considered to supersede #21833 (although technically there's no particular reason that they can't both be implemented).

**Comment by alexrp** [comment]
cc @rohlem @Rexicon226 @dweiller @ParfenovIgor since you participated in the discussion on #21833.

**Comment by rohlem** [comment]
I like the idea, and this mechanism seems like the right tool for the job.

The one detail I'm not 100% sure on is the current interplay with inlining (and I think this isn't fully addressed in the langref either):

> This builtin function would work similarly to `@setRuntimeSafety()` in that it affects the surrounding scope.

```zig
fn regular_f() void {} //code
noinline fn noinline_f() void {} //code
inline fn inline_f() void {} //code
fn main() { //scope 0
  @setRuntimeSafety(x);
  @setIntrinsicRecognition(y);
  { //scope 1
    // code
  }
  noinline_f();
  inline_f();
  regular_f();
}
```
The [current example in the langref](https://ziglang.org/documentation/master/#setRuntimeSafety) makes it seem like the settings of scope 0 transitively apply to nested scopes (scope 1 here).
I think it would be highly surprising if `noinline` functions, or function called with `.never_inline` modifier were affected by the caller scope's settings.

What seems unclear to me is whether an `inline` function (or a call with `.always_inline`) would keep the caller's settings (as if it were directly inlined as a scope), or reset to the defaults.
If it does keep the settings, then what about `regular_fn`? In particular, could the compiler's decision on whether or not to inline a particular function call affect that function's runtime safety and intrinsic recognition settings? That would seem very surprising to me.

I think the simplest (most regular) answer would be to **reset these scope-local settings for all functions, even for `inline fn`**.
(If we *do* want to support all combinations, maybe `@call`'s `CallModifier` enum could be expanded into a `CallOptions` struct, allowing a caller to specify the callee's default runtime safety and intrinsic recognition settings.)

**Comment by alexrp** [comment]
> The one detail I'm not 100% sure on is the current interplay with inlining (and I think this isn't fully addressed in the langref either):

Whatever we decide here, it just needs to be consistent among these scoped `@set*()` builtin functions.

> I think the simplest (most regular and least surprising) answer would be to **reset these scope-local settings for all functions, even for `inline fn`**.

I think I'd agree with this.

**Comment by alexrp** [comment]
Hmm... one important thing to note here is that LLVM only allows `no-builtins` at the function level. This effectively means that any `@setIntrinsicRecognition(false)` inside a function would win out over any `@setIntrinsicRecognition(true)` (unless we do outlining, but that's crazy complicated and definitely not worth it just for this).

This definitely calls into question the value of having `@setIntrinsicRecognition()` be defined as scoped rather than be defined as a property of the function it's invoked in.

**Comment by dweiller** [comment]
I definitely agree with having `@set*` builtins not affect the caller of inline functions makes sense.

As far as `@setIntrinsicRecognition` goes I agree that it would be good to be able to control recognition without having to affect a whole module. If we have `@setInstrinsicRecognition` I would even think it makes sense to remove `-fno-builtin` and just require calls to `@setIntrinsicRecognition` in code that needs it. It feels too much like two ways to do the same thing to me, and applying it to whole modules could lead to missed optimisation opportunities for functions that didn't need it.

> This definitely calls into question the value of having `@setIntrinsicRecognition()` be defined as scoped rather than be defined as a property of the function it's invoked in.

I think it could be fine to still have it defined at scope-level and let backends that can't support that level of granularity could just hoist `@setIntrinsicRecognition(false)` up to whatever scope they support. On the other hand, I'm not sure what kind of code would want different settings at different scopes in the same function - the only use-case I have in mind is preventing recursive function calls. Maybe cases where you could get mutually recursive function calls between intrinsics is where you'd want sub-function-scope targetting?

**Comment by alexrp** [comment]
> If we have `@setInstrinsicRecognition` I would even think it makes sense to remove `-fno-builtin` and just require calls to `@setIntrinsicRecognition` in code that needs it. It feels too much like two ways to do the same thing to me, and applying it to whole modules could lead to missed optimisation opportunities for functions that didn't need it.

That's fine by me. We'd need to carefully audit compiler-rt when removing `-fno-builtin`, but that's doable.

> I think it could be fine to still have it defined at scope-level and let backends that can't support that level of granularity could just hoist `@setIntrinsicRecognition(false)` up to whatever scope they support. On the other hand, I'm not sure what kind of code would want different settings at different scopes in the same function - the only use-case I have in mind is preventing recursive function calls. Maybe cases where you could get mutually recursive function calls between intrinsics is where you'd want sub-function-scope targetting?

I can't think of any use cases either.

It's *probably* fine for any `@setIntrinsicRecognition(false)` to win out. I can't actually think of any cases where it would be sensible to actively require `@setIntrinsicRecognition(true)` to kick in for correctness. Such code would be incredibly fragile (depends on compiler version, build mode, etc...).

**Comment by alexrp** [comment]
I changed my mind on scoped vs function-level; it turns out to be much easier to implement at the function level, and also, that's how `@disableInstrumentation()` already works, so there's precedent.

**Comment by alexrp** [comment]
Proposal text updated based on the implementation PR (#22154).

**Comment by wooster0** [comment]
I think this is another case where this builtin would be useful: https://github.com/ziglang/zig/pull/22854

**Comment by dweiller** [comment]
It looks like a `setIntrinsicRecognition(false)` call just turning on the 'no-builtins' attribute would not be enough for the LLVM backend. We would also have to ensure that calls to problematic llvm builtins are not produced by the llvm backend. For example when lowering compiler-rt `memcpy` the LLVM backend needs to not produce calls to `llvm.memcpy`, but I believe `llvm.memcpy.inline` should be fine. This would mean that the backend could just pick which LLVM builtin is used based on the 'intrinsic recognition' setting.

**Comment by dweiller** [comment]
Yes, that makes sense - I just wanted to have it mentioned somewhere on the issue tracker.

**Comment by andrewrk** [comment]
> Consider one motivating use case from the standard library:
> (posixCallMainAndExit)
> (pie relocate)

Just want to note that these are not motivating use cases, since calling into intrinsics such as memcpy would actually be good. Such things would be built in a different compilation unit and not subject to the problem that `@disableInstrumentation()` is trying to solve.

As far as I can tell, the only motivating use case for these is compiler-rt itself, however, I think that is compelling enough alone to warrant the language feature.

**Comment by alexrp** [comment]
> Just want to note that these are not motivating use cases, since calling into intrinsics such as memcpy would actually be good. Such things would be built in a different compilation unit and not subject to the problem that `@disableInstrumentation()` is trying to solve.

But they are motivating use cases: Because PIE relocations have not been applied yet, we're not allowed to make calls at all because they go through the global pointer/GOT register on some targets (for some targets, even in the same compilation unit!). So we can't have the compiler turning e.g. large load/store operations or copy loops into `memcpy`/`memmove` calls.

**Comment by andrewrk** [comment]
I see, thanks for elaborating 👍
